pkg/edgehub/controller_test.go

```diff
@@ -0,0 +1,820 @@
+				t.Errorf("NewEdgeHubController() KeeperLock = %v, want %v", got.keeperLock, tt.want.keeperLock)
+			}
+			if reflect.TypeOf(got.stopChan) != reflect.TypeOf(tt.want.stopChan) {
+				t.Errorf("NewEdgeHubController() StopChan = %v, want %v", got.stopChan, tt.want.stopChan)
+			}
+			if !reflect.DeepEqual(got.syncKeeper, tt.want.syncKeeper) {
+				t.Errorf("NewEdgeHubController() SyncKeeper = %v, want %v", got.syncKeeper, tt.want.syncKeeper)
+			}
+		})
+	}
+}
+
+//TestInitial() tests the procurement of the cloudhub client
+func TestInitial(t *testing.T) {
+	controllerConfig := config.ControllerConfig{
+		HeartbeatPeriod: 150 * time.Second,
+		RefreshInterval: 15 * time.Minute,
+		AuthInfosPath:   "/var/IEF/secret",
+		PlacementURL:    TestServer.URL + "/proper_request",
+		ProjectID:       "foo",
+		NodeID:          "bar",
+	}
+	tests := []struct {
+		name             string
+		fields           Controller
+		webSocketConfig  config.WebSocketConfig
+		controllerConfig config.ControllerConfig
+		ctx              *context.Context
+		expectedError    error
+	}{
+		{"TestEdgeHubController_initial: Proper input", Controller{
+			config:     &controllerConfig,
+			stopChan:   make(chan struct{}),
+			syncKeeper: make(map[string]chan model.Message),
+		}, config.WebSocketConfig{
+			CertFilePath: CertFile,
+			KeyFilePath:  KeyFile,
+		}, config.ControllerConfig{
+			PlacementURL: TestServer.URL,
+			ProjectID:    "foo",
+			NodeID:       "bar",
+		},
+			context.GetContext(context.MsgCtxTypeChannel), nil},
+
+		{"TestEdgeHubController_initial: Wrong placement URL", Controller{
+			config:     &controllerConfig,
+			stopChan:   make(chan struct{}),
+			syncKeeper: make(map[string]chan model.Message),
+		}, config.WebSocketConfig{
+			CertFilePath: CertFile,
+			KeyFilePath:  KeyFile,
+		}, config.ControllerConfig{
+			PlacementURL: TestServer.URL,
+			ProjectID:    "foo",
+			NodeID:       "bar",
+		},
+			context.GetContext(context.MsgCtxTypeChannel), nil},
+
+		{"TestEdgeHubController_initial: No project Id & node Id", Controller{
+			config:     &controllerConfig,
+			stopChan:   make(chan struct{}),
+			syncKeeper: make(map[string]chan model.Message),
+		}, config.WebSocketConfig{
+			CertFilePath: CertFile,
+			KeyFilePath:  KeyFile,
+		}, config.ControllerConfig{
+			PlacementURL: TestServer.URL,
+			ProjectID:    "",
+			NodeID:       "",
+		},
+			context.GetContext(context.MsgCtxTypeChannel), nil},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			edgeHubConfig := config.GetConfig()
+			edgeHubConfig.WSConfig = tt.webSocketConfig
+			edgeHubConfig.CtrConfig = tt.controllerConfig
+			ehc := &Controller{
+				context:    tt.fields.context,
+				chClient:   tt.fields.chClient,
+				config:     tt.fields.config,
+				stopChan:   tt.fields.stopChan,
+				syncKeeper: tt.fields.syncKeeper,
+				keeperLock: tt.fields.keeperLock,
+			}
+			if err := ehc.initial(tt.ctx); err != tt.expectedError {
```

Comment:
Is it possible to add a negative test case here ?

Reply:
@sids-b The source code for initial() has been written in such a way that it is not possible to add a negative test case here